Source/x11/XGServerWindow.m

```diff
@@ -4409,24 +4413,76 @@ - (void) freecursor: (void*) cid
   return depth;
 }
 
+/* This method assumes that we deal with one X11 screen - `defScreen`.
+   Basically it means that we have DISPLAY variable set to `:0.0`.
+   AppKit and X11 use the same term "screen" with different meaning:
+   AppKit Screen - single mintor/display device.
+   X11 Screen - its a plane where X Server can draw on.
+   XRandR - Xorg extension that halps to manipulate monitors layout providing
+   X11 Screen. */
+#ifdef HAVE_XRANDR
 - (NSArray *)screenList
 {
-  /* I guess screen numbers are in order starting from zero, but we
-     put the main screen first */
- int i;
- int count = ScreenCount(dpy);
- NSMutableArray *screens = [NSMutableArray arrayWithCapacity: count];
- if (count > 0)
-   [screens addObject: [NSNumber numberWithInt: defScreen]];
- for (i = 0; i < count; i++)
-   {
-     if (i != defScreen)
-       [screens addObject: [NSNumber numberWithInt: i]];
-   }
- return screens;
+  NSMutableArray      *screens;
+  int                 i;
+  XRRScreenResources  *screen_res;
+  XRROutputInfo       *output_info;
+  XRRCrtcInfo         *crtc_info;
+
+  screen_res = XRRGetScreenResources(dpy, RootWindow(dpy, defScreen));
+  monitorsCount = screen_res->noutput;
+  
+  screens = [NSMutableArray arrayWithCapacity: monitorsCount];
+  if (monitors != NULL) {
+    NSZoneFree([self zone], monitors);
+  }
+  monitors = NSZoneMalloc([self zone], monitorsCount * sizeof(MonitorDevice));
+
+  for (i = 0; i < monitorsCount; i++)
+    {
+      output_info = XRRGetOutputInfo(dpy, screen_res, screen_res->outputs[i]);
+      if (output_info->crtc)
+        {
+          crtc_info = XRRGetCrtcInfo(dpy, screen_res, output_info->crtc);
+
+          // Fill the cache of device parameters
+          monitors[i].depth = [self windowDepthForScreen: i];
+          monitors[i].resolution = [self resolutionForScreen: i];
+          monitors[i].frame = NSMakeRect(crtc_info->x, crtc_info->y,
+                                         crtc_info->width, crtc_info->height);
+          // Add number of device
+          [screens addObject: [NSNumber numberWithInt: i]];
+        }
+    }
+  
+  XRRFreeScreenResources(screen_res);
+  
+  return screens;
+}
+#else // HAVE_XRANDR
+- (NSArray *)screenList
+{
+  int i;
+  NSMutableArray *screens;
+ 
+  monitorsCount = ScreenCount(dpy);
+  screens = [NSMutableArray arrayWithCapacity: monitorsCount];
+  if (monitorsCount > 0)
+    {
+      /* I guess screen numbers are in order starting from zero, but we
+         put the main screen first */
+      [screens addObject: [NSNumber numberWithInt: defScreen]];
+    }
+  for (i = 0; i < count; i++)
+    {
+      if (i != defScreen)
+        [screens addObject: [NSNumber numberWithInt: i]];
+    }
+  return screens;
 }
+#endif
 
-- (NSWindowDepth) windowDepthForScreen: (int) screen_num
+- (NSWindowDepth) windowDepthForScreen: (int) screen_num // X Screen - OK
```

Comment:
I really hate comments at the end of method definitions. If this is going to stay could you please move it before the method and make an explicit and understandable statement out of it?

Reply:
Actually these are not comments but marks for me. I need to check end fix parameter name to be more clear what this method woud expect X11 screen number or NSScreen index. I'll remove it later.

Reply:
I've finished fixing marked methods and removed marks.